src/providers/base.rs

```diff
@@ -0,0 +1,79 @@
+use anyhow::Result;
+use serde::{Deserialize, Serialize};
+
+use super::types::{message::Message, tool::Tool};
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Usage {
+    pub input_tokens: Option<i32>,
+    pub output_tokens: Option<i32>,
+    pub total_tokens: Option<i32>,
+}
+
+impl Usage {
+    pub fn new(
+        input_tokens: Option<i32>,
+        output_tokens: Option<i32>,
+        total_tokens: Option<i32>,
+    ) -> Self {
+        Self {
+            input_tokens,
+            output_tokens,
+            total_tokens,
+        }
+    }
+}
+
+/// Base trait for AI providers (OpenAI, Anthropic, etc)
+pub trait Provider: Send + Sync {
+    /// Create a provider instance from environment variables
+    fn from_env() -> Result<Self>
+    where
+        Self: Sized;
+
+    /// Generate the next message using the specified model and other parameters
+    fn complete(
```

Comment:
nit: open question to me in this new version where everything should sit. I see three things here - 
1. connection/authorization - it's in the provider config
2. parametrization of the completion - args in complete like model, temperature
3. the messages - arg in complete

I kind of think 2 actually could go in the provider instantiation. That for one helps avoid only supporting parametrization that not all the various apis support (e.g. do they all support stop_sequences etc). Also I don't think in typical usage you'll be change 

Can think more about this